gpio_emul: Fix bug with input & output pins

jeremybettis opened the pull request:
Change gpio_emul_input_set_masked_pend to log when it returns an error.
Changed all calls to gpio_emul_input_set_masked* to check the return
value.

Added a test case that sets and reads emulated 2 gpios to make sure
they don't interact with each other. Fails before this cl, passes
afterwards.

./scripts/twister -T tests/drivers/gpio

The function gpio_emul_input_set_masked_pend will not set the input
values if there are bits set in the value that are not in the mask. In
almost every call to that function, the value wasn't being limited to
bits set in the mask, so they were failing if there was more than one
gpio configured for INPUT|OUTPUT with a value of 1.

jeremybettis committed:
native_posix: Add gpio to supported list

Add gpio to supported list in native_posix boards so that the gpio tests
with the gpio_emul will run in twister.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis committed:
gpio_emul: Fix bug with input & output pins

Change gpio_emul_input_set_masked_pend to log when it returns an error.
Changed all calls to gpio_emul_input_set_masked* to check the return
value.

Changed the function gpio_emul_input_set_masked_pend to apply the mask
to the provided values and not return an error if there are excess
values. In almost every call to that function, the value wasn't being
limited to bits set in the mask, so they were failing if there was more
than one gpio configured for INPUT|OUTPUT with a value of 1.

Fixes #40646

Signed-off-by: Jeremy Bettis <jbettis@google.com>

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -294,17 +294,23 @@ int gpio_emul_input_set_masked_pend(const struct device *port, gpio_port_pins_t
 	}
 
 	if (~config->common.port_pin_mask & mask) {
+		LOG_ERR("%s: Pin not supported port_pin_mask=%x mask=%x", __func__,
```
Remove

jeremybettis replied:
This function is only callable from test code. If it returns an error, don't you want to know why?

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -294,17 +294,23 @@ int gpio_emul_input_set_masked_pend(const struct device *port, gpio_port_pins_t
 	}
 
 	if (values & ~mask) {
+		LOG_ERR("%s: Extra bits in values=%x mask=%x", __func__, values,
```
Remove

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -294,17 +294,23 @@ int gpio_emul_input_set_masked_pend(const struct device *port, gpio_port_pins_t
 
 	input_mask = get_input_pins(port);
 	if (~input_mask & mask) {
+		LOG_ERR("%s: Not input pin input_mask=%x mask=%x", __func__,
```
Remove

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -412,20 +419,36 @@ static int gpio_emul_pin_configure(const struct device *port, gpio_pin_t pin,
 			if (flags & GPIO_INPUT) {
 				/* for push-pull mode to generate interrupts */
-				gpio_emul_input_set_masked_pend(port, BIT(pin), drv_data->output_vals, false);
+				rv = gpio_emul_input_set_masked_pend(
```
Can you please clarify what exactly this is fixing?

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -412,20 +419,36 @@ static int gpio_emul_pin_configure(const struct device *port, gpio_pin_t pin,
 			if (flags & GPIO_INPUT) {
 				/* for push-pull mode to generate interrupts */
-				gpio_emul_input_set_masked_pend(port, BIT(pin), drv_data->output_vals, false);
+				rv = gpio_emul_input_set_masked_pend(
```
Can you please clarify what this is fixing?

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -412,20 +419,36 @@ static int gpio_emul_pin_configure(const struct device *port, gpio_pin_t pin,
 	} else if (flags & GPIO_INPUT) {
 		if (flags & GPIO_PULL_UP) {
-			gpio_emul_input_set_masked_pend(port, BIT(pin), BIT(pin), false);
+			rv = gpio_emul_input_set_masked_pend(port, BIT(pin), BIT(pin),
```
What is this fixing

cfriedt reviewed (changes requested):
Can you please remove the added log messages and whitespace changes? It would be easier to review if the change only reflected what the minimal necessary changes.

Additionally, why was it necessary to modify the test suite?

cfriedt commented:
> The function gpio_emul_input_set_masked_pend will not set the input values if there are bits set in the value that are not in the mask.

My originally implementation operated the way you describe but I was asked to change it so that it would set bits regardless of what the mask was, because it reflected more closely what actual hardware does.

Generally, GPIO-oriented registers may return values that fall outside of the provided mask. E.g. an application is perfectly OK to use a mask of (-1) to refer to all pins, regardless of which are actually in use, and that should not trigger an error.

cc: @pabigot

jeremybettis commented:
> > The function gpio_emul_input_set_masked_pend will not set the input values if there are bits set in the value that are not in the mask.
> 
> My originally implementation operated the way you describe but I was asked to change it so that it would set bits regardless of what the mask was, because it reflected more closely what actual hardware does.
> 
> Generally, GPIO-oriented registers may return values that fall outside of the provided mask. E.g. an application is perfectly OK to use a mask of (-1) to refer to all pins, regardless of which are actually in use, and that should not trigger an error.
> 
> So currently, I do not believe this change is correct or required.

That is why I added the new test, which fails without my changes. The test demonstrates that the code does not work as intended.

If you called gpio_emul_input_set_masked_pend with a mask of -1 and a value of 3, that would work fine. However if you call it with a mask of 3 and a value of -1, it will return an error, which is exactly what the code in 

> Can you please remove the added log messages and whitespace changes? It would be easier to review if the change only reflected what the minimal necessary changes.

There are no whitespace changes, so I don't understand what you mean.  The only "whitespace" changes are where the line was <80 chars, I added text making it over 80 chars, and so I wrapped it. These are the minimal necessary changes.

> 
> Additionally, why was it necessary to modify the test suite?

I added a new test to prove that the code was not working correctly, and the test passes after my changes to prove that they work.

pabigot commented:
> However if you call it with a mask of 3 and a value of -1, it will return an error, which is exactly what the code in

I'm not active with Zephyr, but my recollection was the design was intended to allow -1 to be passed as a mask meaning "all pins, even if some are missing" but also as a value meaning "set all pins even if they're outside of the mask".  So it sounds to me like the code is wrong.

jeremybettis commented:
> I'm not active with Zephyr, but my recollection was the design was intended to allow -1 to be passed as a mask meaning "all pins, even if some are missing" but also as a value meaning "set all pins even if they're outside of the mask". So it sounds to me like the code is wrong.

So would you suggest completely removing the error return at line 302?

	if (values & ~mask) {
		return -EINVAL;
	}

pabigot commented:
I don't know.  At the time I was working GPIO, that would be the wrong thing to do.  Other GPIO drivers from that time wouldn't treat that as an error.  What do they do now?  An emulator should generally behave like others; if real hardware rejects the combination so should the emulator.

Though IIRC the "emulator" infrastructure is a little unusual, so if this is in a function that is invoked by emulator-specific code that instructs the emulator how to behave, rather than in code that's reached using GPIO driver functions, an error might be appropriate.

I defer to @cfriedt to figure out the best overall approach.

jeremybettis commented:
> Though IIRC the "emulator" infrastructure is a little unusual, so if this is in a function that is invoked by emulator-specific code that instructs the emulator how to behave, rather than in code that's reached using GPIO driver functions, an error might be appropriate.

In this case, my problems are in emulator code calling other emulator code and passing data which causes errors and ignoring those error codes. So my change ignores whether it is proper to return errors and always passes data that won't cause errors instead. I also changed the code to not ignore the errors, in case something like this comes up again.

cfriedt commented:
> I defer to @cfriedt to figure out the best overall approach.

A good test would be to compare against a few different hardware platforms and report results here.

Also, it would be great if @mnkp could weigh in as the current GPIO maintainer.

Ultimately, `gpio_emul` must be consistent with non-emulated hardware.

jeremybettis commented:
> Ultimately, `gpio_emul` must be consistent with non-emulated hardware.

I can guarantee there is no hardware that lets you set one output gpio and read the value back, but if you set 2 output gpios, the 2nd one reads as 0.  That is the current behavior of the emulator.

mnkp reviewed (changes requested):
I believe that a proper fix to the described issue is to change line 468 in `gpio_emul_port_set_masked_raw` to
```
	values &= mask;
```
to make sure that the function will modify only the pins that it is allowed to modify.

Additionally in function `gpio_emul_input_set_masked_pend` we should remove
```
	if (values & ~mask) {
		return -EINVAL;
	}
```
and change line 314 to
```
	drv_data->input_vals |= values & mask;
```
Otherwise in the function `gpio_emul_port_set_masked_raw` the following call
```
	/* in push-pull, set input values & fire interrupts */
	gpio_emul_input_set_masked(port, mask & get_input_pins(port), drv_data->output_vals);
```
will not work as intended.

Regarding the comment
> > The function gpio_emul_input_set_masked_pend will not set the input values if there are bits set in the value that are not in the mask.
> 
> My originally implementation operated the way you describe but I was asked to change it so that it would set bits regardless of what the mask was, because it reflected more closely what actual hardware does.

I am a bit confused. The first statement
> The function gpio_emul_input_set_masked_pend will not set the input values if there are bits set in the value that are not in the mask.

describes the current situation since we have
```
	if (values & ~mask) {
		return -EINVAL;
	}
```

jeremybettis commented:
@mnkp I made the changes you requested. Please take another look. The test still passes.

jeremybettis commented on drivers/gpio/gpio_emul.c:
```diff
@@ -502,13 +533,17 @@ static int gpio_emul_port_clear_bits_raw(const struct device *port,
 	drv_data->output_vals &= ~pins;
 	/* in push-pull, set input values & fire interrupts */
-	gpio_emul_input_set_masked(port, pins & get_input_pins(port), drv_data->output_vals);
+	rv = gpio_emul_input_set_masked(port, pins & get_input_pins(port), drv_data->output_vals);
 	k_mutex_unlock(&drv_data->mu);
+	if (rv != 0) {
+		return rv;
+	}
```
Since ignoring error codes is what got us into this situation, I'm happy to convert them to asserts.

jeremybettis commented on drivers/gpio/gpio_emul.c:
```diff
@@ -465,12 +475,14 @@ static int gpio_emul_port_set_masked_raw(const struct device *port,
 	drv_data->output_vals |= values;
 	/* in push-pull, set input values & fire interrupts */
-	gpio_emul_input_set_masked(port, mask & get_input_pins(port), drv_data->output_vals);
+	__ASSERT_NO_MSG(!gpio_emul_input_set_masked(port,
```
I'm sorry, I misunderstood. I've put it back to the previous version.

Should I just put the asserts into gpio_emul_input_set_masked_pend instead, so they only have to go in one place? Or are the asserts just not worth the effort?

mnkp replied:
> Should I just put the asserts into gpio_emul_input_set_masked_pend instead, so they only have to go in one place?

The function can be called directly by the user code, not only by the gpio_emul driver itself so we cannot place ASSERT directly in `gpio_emul_input_set_masked_pend`.

> Or are the asserts just not worth the effort?

ASSERTS are generally a very useful tool, however in this specific case they require a fair amount of extra code that overall make the driver more difficult to read. I personally wouldn't add it but if you prefer to check the return value of the function call I'm not going to block it.

What I meant is that we should keep the ASSERT and use `rv` variable at the same time. I.e. in your latest version you should replace all occurrences of
```

	if (rv != 0) {
		return rv;
	}
```
with
```
	__ASSERT_NO_MSG(rv == 0);
```
If you want to keep the ASSERT.

jeremybettis replied:
I don't need to keep the asserts, I think it is fine as is.

mnkp replied:
By design a call to `gpio_emul_input_set_masked` from withing the driver itself must always return 0. Any other value would mean there is a bug in the driver code. The check of return value is not necessary. We could add an ASSERT to verify that our assumption is correct but as I mentioned for me that's not necessary.

jeremybettis replied:
I see what you are saying now, you don't want internal calls to ever cause bad return values into non-test code. Changed.

To verify that there are no compiler errors, I did these steps:

* I added an unused variable to gpio_emul_pin_configure and verified that `./scripts/twister -T tests/drivers/gpio` failed.
* I added `CONFIG_FORCE_NO_ASSERT=y` to `zephyr/tests/drivers/gpio/gpio_api_1pin/prj.conf` and verified that `./scripts/twister -T tests/drivers/gpio` passed.
* I added `CONFIG_ASSERT=y` and `CONFIG_ASSERT_LEVEL=1` to `zephyr/tests/drivers/gpio/gpio_api_1pin/prj.conf`, and commented out the #warning in `include/sys/__assert.h` and verified that `./scripts/twister -T tests/drivers/gpio` passed.

jeremybettis commented:
> Can you please remove the added log messages and whitespace changes? It would be easier to review if the change only reflected what the minimal necessary changes.
> 
> Additionally, why was it necessary to modify the test suite?

@cfriedt Can you please resolve this?

cfriedt commented:
Just for clarification, the title of this PR is inaccurate, as the existing previous behaviour was correct based on the previous requirements. I'm fine with those requirements evolving though, as long as nothing breaks and it is still useful.

jeremybettis commented:
> Just for clarification, the title of this PR is inaccurate, as the existing previous behaviour was correct based on the previous requirements (i.e. there was no "bug"). I'm fine with those requirements evolving though, as long as nothing breaks and it is still useful.

There was a bug, in that calling gpio_emul_driver->port_set_bits_raw() to enable a OUTPUT|INPUT pin, followed by gpio_emul_driver->port_get_raw() to read that same pin would return 0 for the value if there were any other GPIO output pins also set to 1.

cfriedt commented:
> There was a bug, in that calling gpio_emul_driver->port_set_bits_raw() to enable a OUTPUT|INPUT pin, followed by gpio_emul_driver->port_get_raw() to read that same pin would return 0 for the value if there were any other GPIO output pins also set to 1.

Thanks for clarifying. That does sound like a bug. Please create an issue to link against this PR, and update your commit to say `Fixes #[issue]`.

I'll add the appropriate labels.

jeremybettis commented:
>  Please create an issue to link against this PR, and update your commit to say `Fixes #[issue]`.

done

cfriedt commented:
Fixes #40646
